frame/im-online/src/lib.rs

```diff
@@ -129,37 +130,64 @@ pub mod ed25519 {
 	pub type AuthorityId = app_ed25519::Public;
 }
 
-/// The local storage database key under which the worker progress status
-/// is tracked.
-const DB_KEY: &[u8] = b"parity/im-online-worker-status";
+const DB_PREFIX: &[u8] = b"parity/im-online-heartbeat/";
+/// How many blocks do we wait for heartbeat transaction to be included
+/// before sending another one.
+const INCLUDE_THRESHOLD: u32 = 3;
 
-/// It's important to persist the worker state, since e.g. the
-/// server could be restarted while starting the gossip process, but before
-/// finishing it. With every execution of the off-chain worker we check
-/// if we need to recover and resume gossipping or if there is already
-/// another off-chain worker in the process of gossipping.
+/// Status of the offchain worker code.
+///
+/// This stores the block number at which heartbeat was requested and when the worker
+/// has actaully managed to produce it.
+/// Note we store such status for every `authority_index` separately.
 #[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
-struct WorkerStatus<BlockNumber> {
-	done: bool,
-	gossipping_at: BlockNumber,
+struct HeartbeatStatus<BlockNumber> {
+	/// A block number at which we were supposed to send the heartbeat.
+	///
+	/// I.e. the block that was set as the threshold by `SessionHandler` via `HeartbeatAt`.
+	pub heartbeat_at: BlockNumber,
+	/// A block number at which the heartbeat was actually sent.
+	///
+	/// Note that this may be 0 in case sending failed.
+	pub sent_at: BlockNumber,
+}
+
+impl<BlockNumber: PartialEq + SimpleArithmetic + Copy> HeartbeatStatus<BlockNumber> {
+	/// Returns true if heartbeat has been recently sent.
+	///
+	/// Note this function will return `true` iff:
+	/// 1. the requested block number (`heartbeat_at`) is the same
+	/// 2. the heartbeat has been sent recently
+	fn is_recent(&self, heartbeat_at: BlockNumber, now: BlockNumber) -> bool {
```

Comment:
could do with a description of the parameters as it's not immediately obvious what `heartbeat_at` means in the context of repeated heartbeat sending attempts.